libyul/AsmParser.cpp

```diff
@@ -512,14 +512,24 @@ variant<Literal, Identifier> Parser::parseLiteralOrIdentifier()
 	case Token::FalseLiteral:
 	{
 		LiteralKind kind = LiteralKind::Number;
+		string currLiteral = currentLiteral();
+
 		switch (currentToken())
 		{
 		case Token::StringLiteral:
 		case Token::HexStringLiteral:
 			kind = LiteralKind::String;
 			break;
 		case Token::Number:
-			if (!isValidNumberLiteral(currentLiteral()))
+			currLiteral.erase(std::remove(currLiteral.begin(), currLiteral.end(), '_'), currLiteral.end());
+			if (currLiteral.find("e") != string::npos && !(boost::starts_with(currLiteral, "0x") || boost::starts_with(currLiteral, "0X")))
+			{
+				istringstream os(currLiteral);
+				double d;
+				os >> d;
+				currLiteral = to_string(d).substr(0, to_string(d).find('.'));
+			}
```

Comment:
```suggestion
```
This is completely unnecessary, and makes check for things that are not in the scope of this task - this is also reflected in the test cases.

Reply:
Hi @nikola-matic, the issue has a request for both underscores and scientific notation: 

"Currently within assembly blocks, it is not allowed to write a number like `1e6` or with underscores. However, these kinds of notations make it easy to read a number." 

I can take out the scientific notation functionality but I thought it was a good idea since number literals also support scientific notation. This would add an amount of consistency. Let me know what you think?

Reply:
Ah, scientific notation wasn't in the example posted in the issue, so didn't even notice. Non the less, the underscore separated literals are in my opinion simpler than scientific notation, and should none the less be done in separate pull requests (and should probably have separate issues as well), so I would still prefer to keep this PR focused only on underscore separators. None the less, let's wait until tomorrow, and we'll see what @ekpyron thinks.

Reply:
@ekpyron Can you take a look at this and let us know if you want this as 2 PRs or one please?

Reply:
See my opinion in https://github.com/ethereum/solidity/issues/13407#issuecomment-1398259752.

I think that support for underscores is a nice enhancement, especially given that you cannot do arithmetic on constants in Yul. Scientific notation on the other hand makes the most sense when you're dealing with fractional numbers and we don't have these in Yul. I'd hold off on implementing that until we have a clear decision in favor of that in the issue. So I agree with @nikola-matic that this PR should focus only on the underscores.

Reply:
I'm not convinced that we even want underscore support in Yul.